### PR TITLE
Update to RxJava2

### PR DESCRIPTION
**Pull request by apoi**

### CODE REVIEW

**Comment by apoi** [comment]
Ping @tomaszpolanski :)

File: app/src/androidTest/java/io/reark/rxgithubapp/advanced/data/stores/GitHubRepositoryStoreTest.java [hunk 1]
```diff
@@ -220,14 +216,14 @@ public void getOnceAndStream_WithInitialValue_WithDelayedSubscription_ReturnsFir
         gitHubRepositoryStore.put(value2);
 
         // Subscribe to stream that was created potentially a long time ago.
-        stream.subscribe(testSubscriber);
+        stream.subscribeWith(testObserver);
```
Comment (tomaszpolanski) [inline app/src/androidTest/java/io/reark/rxgithubapp/advanced/data/stores/GitHubRepositoryStoreTest.java]: Why not just the plain `test()`?

File: app/src/main/java/io/reark/rxgithubapp/advanced/data/stores/UserSettingsStore.java [hunk 2]
```diff
@@ -50,7 +50,7 @@ public UserSettingsStore(final int userId, @NonNull final ContentResolver conten
 
     private void initUserSettings(final int userId) {
         getOnce(userId)
-                .toObservable()
+                .toMaybe()
```
Comment (tomaszpolanski) [inline app/src/main/java/io/reark/rxgithubapp/advanced/data/stores/UserSettingsStore.java]: Is this needed? Removing it should work as well

File: appshared/src/main/java/io/reark/rxgithubapp/shared/view/RepositoriesView.java [hunk 3]
```diff
@@ -143,11 +142,8 @@ protected void bindInternal(@NonNull final CompositeSubscription s) {
                     .observeOn(AndroidSchedulers.mainThread())
                     .subscribe(viewModel::setSearchString));
             s.add(Observable.create(
-                    subscriber -> {
-                        view.repositoriesAdapter.setOnClickListener(
-                                this::repositoriesAdapterOnClick);
-                        subscriber.add(Subscriptions.create(() ->
-                                view.repositoriesAdapter.setOnClickListener(null)));
+                    emitter -> {
```
Comment (tomaszpolanski) [inline appshared/src/main/java/io/reark/rxgithubapp/shared/view/RepositoriesView.java]: Should be replaced with:
```
Completable.fromAction(
                    () -> view.repositoriesAdapter.setOnClickListener(this::repositoriesAdapterOnClick)
```

File: appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java [hunk 4]
```diff
@@ -84,17 +84,17 @@ public RepositoriesViewModel(@NonNull final GetGitHubRepositorySearch getGitHubR
 
     @NonNull
     public Observable<GitHubRepository> getSelectRepository() {
-        return selectRepository.asObservable();
+        return selectRepository;
```
Comment (tomaszpolanski) [inline appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java]: Use `.hide()`

File: appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java [hunk 5]
```diff
@@ -84,17 +84,17 @@ public RepositoriesViewModel(@NonNull final GetGitHubRepositorySearch getGitHubR
...
     }
 
     @NonNull
     public Observable<List<GitHubRepository>> getRepositories() {
-        return repositories.asObservable();
+        return repositories;
```
Comment (tomaszpolanski) [inline appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java]: Use `.hide()`

File: appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java [hunk 6]
```diff
@@ -84,17 +84,17 @@ public RepositoriesViewModel(@NonNull final GetGitHubRepositorySearch getGitHubR
...
     }
 
     @NonNull
     public Observable<ProgressStatus> getNetworkRequestStatusText() {
-        return networkRequestStatusText.asObservable();
+        return networkRequestStatusText;
```
Comment (tomaszpolanski) [inline appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java]: Use `.hide()`

File: appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoryViewModel.java [hunk 7]
```diff
@@ -58,20 +58,19 @@ public RepositoryViewModel(@NonNull final GetUserSettings getUserSettings,
...
+                .switchMap(fetchAndGetGitHubRepository::call)
+                .filter(DataStreamNotification::isOnNext)
+                .map(DataStreamNotification::getValue)
+                .subscribe(repository::onNext));
     }
 
     @NonNull
     public Observable<GitHubRepository> getRepository() {
-        return repository.asObservable();
+        return repository;
```
Comment (tomaszpolanski) [inline appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoryViewModel.java]: Use `.hide()`

File: reark/src/main/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBase.java [hunk 8]
```diff
@@ -298,27 +296,31 @@ private U mergedItem(@NonNull final U currentItem, @NonNull final U newItem) {
...
+    private Single<Boolean> createModifyingOperation(@NonNull final Function<Subject<Boolean>, CoreValue<U>> valueFunc) {
         BehaviorSubject<Boolean> completionNotifier = BehaviorSubject.create();
-        operationSubject.onNext(valueFunc.call(completionNotifier));
 
-        return completionNotifier.first()
-                .toSingle();
+        try {
+            operationSubject.onNext(valueFunc.apply(completionNotifier));
+        } catch (Exception e) {
+            Log.e(TAG, "Error creating operation!", e);
```
Comment (tomaszpolanski) [inline reark/src/main/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBase.java]: Shouldn't you push the error to the subject?

File: reark/src/main/java/io/reark/reark/data/stores/cores/MemoryStoreCore.java [hunk 9]
```diff
@@ -131,7 +130,11 @@ public MemoryStoreCore(@NonNull final Func2<U, U, U> putMergeFunction) {
 
             if (!valuesEqual) {
                 Log.v(TAG, "Merging values at " + id);
-                newItem = putMergeFunction.call(currentItem, newItem);
+                try {
+                    newItem = putMergeFunction.apply(currentItem, newItem);
+                } catch (Exception e) {
+                    Log.e(TAG, "Error merging!", e);
```
Comment (tomaszpolanski) [inline reark/src/main/java/io/reark/reark/data/stores/cores/MemoryStoreCore.java]: Shouldn't this go to the `Single` as an error?

File: reark/src/test/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBaseTest.java [hunk 10]
```diff
@@ -29,55 +28,55 @@ public void setup() {
     }
 
     private static Observable<Integer> createSource(int numItems) {
-        return Observable.fromEmitter(publisher -> {
+        return Observable.fromPublisher(publisher -> {
```
Comment (tomaszpolanski) [inline reark/src/test/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBaseTest.java]: Wouldn't `Observable.range(1, numItems)` be better?

File: app/src/androidTest/java/io/reark/rxgithubapp/advanced/data/stores/GitHubRepositoryStoreTest.java (hunk 1 above)
Comment (apoi) [reply app/src/androidTest/java/io/reark/rxgithubapp/advanced/data/stores/GitHubRepositoryStoreTest.java]: Changed to `test()`

File: app/src/main/java/io/reark/rxgithubapp/advanced/data/stores/UserSettingsStore.java (hunk 2 above)
Comment (apoi) [reply app/src/main/java/io/reark/rxgithubapp/advanced/data/stores/UserSettingsStore.java]: Indeed, removed

File: appshared/src/main/java/io/reark/rxgithubapp/shared/view/RepositoriesView.java (hunk 3 above)
Comment (apoi) [reply appshared/src/main/java/io/reark/rxgithubapp/shared/view/RepositoriesView.java]: Done

File: appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java (hunk 4 above)
Comment (apoi) [reply appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java]: Is this useful only for subjects or in some other context too?

File: appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java (hunk 5 above)
Comment (apoi) [reply appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java]: Done

File: appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java (hunk 6 above)
Comment (apoi) [reply appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoriesViewModel.java]: Done

File: appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoryViewModel.java (hunk 7 above)
Comment (apoi) [reply appshared/src/main/java/io/reark/rxgithubapp/shared/viewmodels/RepositoryViewModel.java]: Done

File: reark/src/main/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBase.java (hunk 8 above)
Comment (apoi) [reply reark/src/main/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBase.java]: The apply call can't really throw, the try-catch is just to satisfy the Function method signature. I seem to recall there was a non-throwing version of the Rx functional interfaces but can't now find anything

File: reark/src/main/java/io/reark/reark/data/stores/cores/MemoryStoreCore.java (hunk 9 above)
Comment (apoi) [reply reark/src/main/java/io/reark/reark/data/stores/cores/MemoryStoreCore.java]: Yes it should, done

File: reark/src/test/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBaseTest.java (hunk 10 above)
Comment (apoi) [reply reark/src/test/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBaseTest.java]: It is, fixed

**Comment by tomaszpolanski** [comment]
For me it looks good now

**Pull request merged**